File: ts/eve/encoding/conversation.ts
/**
 * Conversation Encoding
 * 
 * Ported from backend/services/encoding/conversation_encoding.py
 * Handles generic conversation encoding for analysis, embeddings, and retrieval
 */

import { REACTION_EMOJIS } from '../utils/constants.js';

export interface EncodedMessage {
  id: number;
  guid: string;
  timestamp: string; // ISO string
  sender_id: number | null;
  sender_name?: string;
  text?: string;
  content?: string;
  is_from_me?: number | boolean;
  sender_is_me?: number | boolean;
  attachments?: Array<{
    id: number;
    mime_type?: string;
    file_name?: string;
    filename?: string; // Support both spellings
    is_sticker?: number | boolean;
  }>;
  reactions?: Array<{
    reaction_type?: number;
    associated_message_type?: number; // Legacy field name
    sender_id: number;
    sender_name?: string;
    is_from_me?: number | boolean;
  }>;
  conversation_id: number;
  chat_id: number;
}

export interface EncodedConversation {
  id: number;
  chat_id: number;
  start_time: string; // ISO string
  end_time: string; // ISO string
  messages: EncodedMessage[];
}

export interface EncodeOptions {
  includeSender?: boolean;
  includeAttachments?: boolean;
  includeReactions?: boolean;
  includeStartDate?: boolean;
  includeSendTime?: boolean;
}

/**
 * Parse timestamp to Date object with fallbacks
 */
function parseTimestamp(value: any): Date {
  if (value instanceof Date) {
    return value;
  }

  if (typeof value === 'string') {
    let s = value;
    // Handle ISO8601 with 'Z' suffix
    if (s.endsWith('Z')) {
      s = s.slice(0, -1) + '+00:00';
    }
    try {
      return new Date(s);
    } catch {
      // Fallback: try parsing as seconds since epoch
      try {
        return new Date(parseFloat(s) * 1000);
      } catch {
        console.warn(`Unable to parse timestamp '${value}'; using epoch`);
        return new Date(0);
      }
    }
  }

  if (typeof value === 'number') {
    // Assume seconds since epoch
    return new Date(value * 1000);
  }

  console.warn(`Unexpected timestamp type ${typeof value}; using epoch`);
  return new Date(0);
}

/**
 * Format timestamp as readable string
 */
function formatTimestamp(date: Date, includeDate: boolean = false): string {
  if (includeDate) {
    // Format: "Monday Oct 27, 2025 - 3:45pm"
    const options: Intl.DateTimeFormatOptions = {
      weekday: 'long',
      month: 'short',
      day: 'numeric',
      year: 'numeric',
      hour: 'numeric',
      minute: '2-digit',
      hour12: true,
    };
    return new Intl.DateTimeFormat('en-US', options).format(date).replace(/AM|PM/, (m) => m.toLowerCase());
  } else {
    // Format: "3:45pm"
    const options: Intl.DateTimeFormatOptions = {
      hour: 'numeric',
      minute: '2-digit',
      hour12: true,
    };
    return new Intl.DateTimeFormat('en-US', options).format(date).replace(/AM|PM/, (m) => m.toLowerCase());
  }
}

/**
 * Format reactions for a message
 */
function formatReactions(reactions: EncodedMessage['reactions']): string {
  if (!reactions || reactions.length === 0) {
    return '';
  }

  const reactionCounts: Record<string, number> = {};
  
  for (const react of reactions) {
    const rType = react.associated_message_type || react.reaction_type;
    if (rType !== undefined && rType in REACTION_EMOJIS) {
      const emoji = REACTION_EMOJIS[rType];
      reactionCounts[emoji] = (reactionCounts[emoji] || 0) + 1;
    }
  }

  if (Object.keys(reactionCounts).length === 0) {
    return '';
  }

  const parts = Object.entries(reactionCounts).map(([emoji, count]) =>
    count > 1 ? `${emoji}(${count})` : emoji
  );

  return `[${parts.join(', ')}]`;
}

/**
 * Encode a single message into text format
 */
export function encodeMessage(
  message: EncodedMessage,
  options: EncodeOptions = {}
): string {
  const {
    includeSender = true,
    includeAttachments = true,
    includeReactions = true,
    includeSendTime = false,
  } = options;

  const parts: string[] = [];

  // Add timestamp if requested
  if (includeSendTime) {
    const ts = parseTimestamp(message.timestamp);
    const timeStr = formatTimestamp(ts, false);
    parts.push(`[${timeStr}]`);
  }

  // Add sender name
  if (includeSender) {
    const sender = message.sender_name || 'Unknown';
    parts.push(`${sender}:`);
  }

  // Add message text (support both 'text' and 'content' fields)
  const textValue = message.text || message.content;
  if (textValue) {
    parts.push(textValue);
  }

  // Add attachments
  if (includeAttachments && message.attachments && message.attachments.length > 0) {
    const attTexts: string[] = [];
    for (const att of message.attachments) {
      if (att.mime_type?.startsWith('image/')) {
        attTexts.push('[Image]');
      } else {
        const filename = att.filename || att.file_name || 'Unknown file';
        attTexts.push(`[Attachment: ${filename}]`);
      }
    }
    if (attTexts.length > 0) {
      parts.push(attTexts.join(' '));
    }
  }

  // Add reactions
  if (includeReactions && message.reactions) {
    const reactionText = formatReactions(message.reactions);
    if (reactionText) {
      parts.push(reactionText);
    }
  }

  return parts.join(' ');
}

/**
 * Encode a conversation into text format
 */
export function encodeConversation(
  conversation: EncodedConversation,
  options: EncodeOptions = {}
): string {
  const {
    includeSender = true,
    includeAttachments = true,
    includeReactions = true,
    includeStartDate = false,
    includeSendTime = false,
  } = options;

  // Sort messages by timestamp
  const messages = [...conversation.messages].sort((a, b) => {
    const dateA = parseTimestamp(a.timestamp);
    const dateB = parseTimestamp(b.timestamp);
    return dateA.getTime() - dateB.getTime();
  });

  const encodedLines: string[] = [];

  // Optional date header
  if (includeStartDate && messages.length > 0) {
    const startTime = parseTimestamp(messages[0].timestamp);
    const dateHeader = formatTimestamp(startTime, true);
    encodedLines.push(`=== ${dateHeader} ===`);
  }

  // Encode each message
  for (const msg of messages) {
    const encodedMsg = encodeMessage(msg, {
      includeSender,
      includeAttachments,
      includeReactions,
      includeSendTime,
    });
    if (encodedMsg) {
      encodedLines.push(encodedMsg);
    }
  }

  return encodedLines.join('\n');
}

/**
 * Encode multiple conversations into a map of id -> text
 */
export function encodeConversations(
  conversations: EncodedConversation[],
  options: EncodeOptions = {}
): Record<string, string> {
  const result: Record<string, string> = {};

  for (const conv of conversations) {
    result[conv.id.toString()] = encodeConversation(conv, options);
  }

  return result;
}

File: internal/encoding/conversation.go
package encoding

import (
	"database/sql"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	_ "github.com/mattn/go-sqlite3"
)

// Message represents a conversation message
type Message struct {
	ID          int64
	GUID        string
	Timestamp   time.Time
	SenderName  string
	Text        string
	IsFromMe    bool
	Attachments []Attachment
	Reactions   []Reaction
}

// Attachment represents a message attachment
type Attachment struct {
	ID        int64
	MimeType  string
	FileName  string
	IsSticker bool
}

// Reaction represents a reaction to a message
type Reaction struct {
	ReactionType int
	SenderName   string
	IsFromMe     bool
}

// Conversation represents a conversation with messages
type Conversation struct {
	ID        int64
	ChatID    int64
	StartTime time.Time
	EndTime   time.Time
	Messages  []Message
}

// EncodeOptions controls what information is included in the encoded output
type EncodeOptions struct {
	IncludeSender      bool
	IncludeAttachments bool
	IncludeReactions   bool
	IncludeStartDate   bool
	IncludeSendTime    bool
}

// DefaultEncodeOptions returns the default encoding options
func DefaultEncodeOptions() EncodeOptions {
	return EncodeOptions{
		IncludeSender:      true,
		IncludeAttachments: true,
		IncludeReactions:   true,
		IncludeStartDate:   false,
		IncludeSendTime:    false,
	}
}

// EncodeResult contains the encoded transcript and metadata
type EncodeResult struct {
	Success      bool   `json:"success"`
	FilePath     string `json:"file_path,omitempty"`
	TokenCount   int    `json:"token_count"`
	MessageCount int    `json:"message_count"`
	EncodedText  string `json:"encoded_text,omitempty"`
	Error        string `json:"error,omitempty"`
}

// LoadConversation loads a conversation from the database
func LoadConversation(dbPath string, conversationID int64) (*Conversation, error) {
	db, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return nil, fmt.Errorf("failed to open database: %w", err)
	}
	defer db.Close()

	// Load conversation metadata
	var conv Conversation
	conv.ID = conversationID

	// Load messages (simplified query - adjust based on actual schema)
	query := `
		SELECT
			m.ROWID as id,
			m.guid,
			datetime(m.date / 1000000000 + strftime('%s', '2001-01-01'), 'unixepoch') as timestamp,
			COALESCE(c.display_name, h.id, 'Unknown') as sender_name,
			COALESCE(m.text, '') as text,
			m.is_from_me
		FROM message m
		LEFT JOIN handle h ON m.handle_id = h.ROWID
		LEFT JOIN contact c ON h.id = c.phone_number
		WHERE m.ROWID = ?
		ORDER BY m.date ASC
	`

	rows, err := db.Query(query, conversationID)
	if err != nil {
		return nil, fmt.Errorf("failed to query messages: %w", err)
	}
	defer rows.Close()

	for rows.Next() {
		var msg Message
		var timestampStr string
		var isFromMe int

		err := rows.Scan(
			&msg.ID,
			&msg.GUID,
			&timestampStr,
			&msg.SenderName,
			&msg.Text,
			&isFromMe,
		)
		if err != nil {
			return nil, fmt.Errorf("failed to scan message: %w", err)
		}

		msg.IsFromMe = isFromMe == 1
		msg.Timestamp, _ = time.Parse("2006-01-02 15:04:05", timestampStr)
		conv.Messages = append(conv.Messages, msg)
	}

	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("row iteration error: %w", err)
	}

	if len(conv.Messages) == 0 {
		return nil, fmt.Errorf("no messages found for conversation %d", conversationID)
	}

	// Set conversation time bounds
	conv.StartTime = conv.Messages[0].Timestamp
	conv.EndTime = conv.Messages[len(conv.Messages)-1].Timestamp

	return &conv, nil
}

// EncodeMessage encodes a single message into text format
func EncodeMessage(msg Message, opts EncodeOptions) string {
	var parts []string

	// Add timestamp if requested
	if opts.IncludeSendTime {
		timeStr := msg.Timestamp.Format("3:04pm")
		parts = append(parts, fmt.Sprintf("[%s]", timeStr))
	}

	// Add sender name
	if opts.IncludeSender {
		parts = append(parts, fmt.Sprintf("%s:", msg.SenderName))
	}

	// Add message text
	if msg.Text != "" {
		parts = append(parts, msg.Text)
	}

	// Add attachments
	if opts.IncludeAttachments && len(msg.Attachments) > 0 {
		for _, att := range msg.Attachments {
			if strings.HasPrefix(att.MimeType, "image/") {
				parts = append(parts, "[Image]")
			} else {
				fileName := att.FileName
				if fileName == "" {
					fileName = "Unknown file"
				}
				parts = append(parts, fmt.Sprintf("[Attachment: %s]", fileName))
			}
		}
	}

	// Add reactions
	if opts.IncludeReactions && len(msg.Reactions) > 0 {
		reactionCounts := make(map[string]int)
		for _, r := range msg.Reactions {
			emoji := reactionTypeToEmoji(r.ReactionType)
			if emoji != "" {
				reactionCounts[emoji]++
			}
		}

		if len(reactionCounts) > 0 {
			var reactionParts []string
			for emoji, count := range reactionCounts {
				if count > 1 {
					reactionParts = append(reactionParts, fmt.Sprintf("%s(%d)", emoji, count))
				} else {
					reactionParts = append(reactionParts, emoji)
				}
			}
			parts = append(parts, fmt.Sprintf("[%s]", strings.Join(reactionParts, ", ")))
		}
	}

	return strings.Join(parts, " ")
}

// EncodeConversation encodes a conversation into text format
func EncodeConversation(conv *Conversation, opts EncodeOptions) string {
	var lines []string

	// Sort messages by timestamp (should already be sorted, but ensure it)
	messages := make([]Message, len(conv.Messages))
	copy(messages, conv.Messages)
	sort.Slice(messages, func(i, j int) bool {
		return messages[i].Timestamp.Before(messages[j].Timestamp)
	})

	// Optional date header
	if opts.IncludeStartDate && len(messages) > 0 {
		startTime := messages[0].Timestamp
		dateHeader := startTime.Format("Monday Jan 2, 2006 - 3:04pm")
		lines = append(lines, fmt.Sprintf("=== %s ===", dateHeader))
	}

	// Encode each message
	for _, msg := range messages {
		encodedMsg := EncodeMessage(msg, opts)
		if encodedMsg != "" {
			lines = append(lines, encodedMsg)
		}
	}

	return strings.Join(lines, "\n")
}

// EncodeConversationToFile encodes a conversation and writes it to a file
func EncodeConversationToFile(dbPath string, conversationID int64, outputPath string) EncodeResult {
	// Load conversation
	conv, err := LoadConversation(dbPath, conversationID)
	if err != nil {
		return EncodeResult{
			Success: false,
			Error:   fmt.Sprintf("failed to load conversation: %v", err),
		}
	}

	// Encode conversation
	opts := DefaultEncodeOptions()
	encodedText := EncodeConversation(conv, opts)

	// Write to file
	if err := os.WriteFile(outputPath, []byte(encodedText), 0644); err != nil {
		return EncodeResult{
			Success: false,
			Error:   fmt.Sprintf("failed to write file: %v", err),
		}
	}

	// Count tokens (rough estimate: ~4 chars per token)
	tokenCount := len(encodedText) / 4

	return EncodeResult{
		Success:      true,
		FilePath:     outputPath,
		TokenCount:   tokenCount,
		MessageCount: len(conv.Messages),
	}
}

// EncodeConversationToString encodes a conversation and returns it as a string
func EncodeConversationToString(dbPath string, conversationID int64) EncodeResult {
	// Load conversation
	conv, err := LoadConversation(dbPath, conversationID)
	if err != nil {
		return EncodeResult{
			Success: false,
			Error:   fmt.Sprintf("failed to load conversation: %v", err),
		}
	}

	// Encode conversation
	opts := DefaultEncodeOptions()
	encodedText := EncodeConversation(conv, opts)

	// Count tokens (rough estimate: ~4 chars per token)
	tokenCount := len(encodedText) / 4

	return EncodeResult{
		Success:      true,
		EncodedText:  encodedText,
		TokenCount:   tokenCount,
		MessageCount: len(conv.Messages),
	}
}

// reactionTypeToEmoji converts iMessage reaction types to emoji
func reactionTypeToEmoji(reactionType int) string {
	// iMessage reaction types (from iMessage database)
	reactionMap := map[int]string{
		2000: "❤️", // Love
		2001: "👍",  // Like
		2002: "👎",  // Dislike
		2003: "😂",  // Laugh
		2004: "‼️", // Emphasize
		2005: "❓",  // Question
	}

	if emoji, ok := reactionMap[reactionType]; ok {
		return emoji
	}
	return ""
}

// GetDefaultOutputPath returns a default output path for encoded conversation
func GetDefaultOutputPath(conversationID int64) string {
	home, _ := os.UserHomeDir()
	tmpDir := filepath.Join(home, ".config", "eve", "tmp")
	os.MkdirAll(tmpDir, 0755)
	return filepath.Join(tmpDir, fmt.Sprintf("conversation_%d.txt", conversationID))
}
